File: nagasenacs/src/Nagasena/Proc/Grammars/SchemaInformedGrammar.cs
using System.Diagnostics;
using System.Collections.Generic;

using EventCode = Nagasena.Proc.Common.EventCode;
using EventType = Nagasena.Proc.Common.EventType;
using EventTypeList = Nagasena.Proc.Common.EventTypeList;
using GrammarOptions = Nagasena.Proc.Common.GrammarOptions;
using IGrammar = Nagasena.Proc.Common.IGrammar;
using XmlUriConst = Nagasena.Proc.Common.XmlUriConst;
using EXISchema = Nagasena.Schema.EXISchema;
using EXISchemaConst = Nagasena.Schema.EXISchemaConst;

namespace Nagasena.Proc.Grammars {

  public abstract class SchemaInformedGrammar : Grammar {

    protected internal const sbyte ELEMENT_FRAGMENT_STATE_BASE = DOCUMENT_STATE_END + 1;
    public const sbyte ELEMENT_FRAGMENT_STATE_TAG = ELEMENT_FRAGMENT_STATE_BASE;
    public const sbyte ELEMENT_FRAGMENT_STATE_CONTENT = ELEMENT_FRAGMENT_STATE_TAG + 1;
    public const sbyte ELEMENT_FRAGMENT_EMPTY_STATE_TAG = ELEMENT_FRAGMENT_STATE_CONTENT + 1;
    public const sbyte ELEMENT_FRAGMENT_EMPTY_STATE_CONTENT = ELEMENT_FRAGMENT_EMPTY_STATE_TAG + 1;

    protected internal SchemaInformedGrammar(sbyte grammarType, GrammarCache stateCache) : base(grammarType, stateCache) {
    }

    public override sealed bool SchemaInformed {
      get {
        return true;
      }
    }

    /// <summary>
    /// Create an EventType instance of item type EventType.ITEM_SCHEMA_AT_INVALID_VALUE.
    /// </summary>
    protected internal EventType createEventTypeSchemaAttributeInvalid(EventType eventTypeSchemaAttribute, EventTypeList eventTypeList) {
      return new EventType(eventTypeSchemaAttribute.uri, eventTypeSchemaAttribute.name, eventTypeSchemaAttribute.URIId, eventTypeSchemaAttribute.NameId, EventCode.EVENT_CODE_DEPTH_THREE, eventTypeList, EventType.ITEM_SCHEMA_AT_INVALID_VALUE, eventTypeSchemaAttribute.subsequentGrammar);
    }

    private EventType createEventTypeXsiNil(EventTypeList eventTypeList, IGrammar subsequentGrammar) {
      return new EventType(XmlUriConst.W3C_2001_XMLSCHEMA_INSTANCE_URI, "nil", XmlUriConst.W3C_2001_XMLSCHEMA_INSTANCE_URI_ID, EXISchemaConst.XSI_LOCALNAME_NIL_ID, EventCode.EVENT_CODE_DEPTH_TWO, eventTypeList, EventType.ITEM_SCHEMA_NIL, subsequentGrammar);
    }

    private EventType createEventTypeXsiType(EventTypeList eventTypeList) {
      return new EventType(XmlUriConst.W3C_2001_XMLSCHEMA_INSTANCE_URI, "type", XmlUriConst.W3C_2001_XMLSCHEMA_INSTANCE_URI_ID, EXISchemaConst.XSI_LOCALNAME_TYPE_ID, EventCode.EVENT_CODE_DEPTH_TWO, eventTypeList, EventType.ITEM_SCHEMA_TYPE, (IGrammar)null);
    }

    ///////////////////////////////////////////////////////////////////////////
    /// Methods computing EventCodeTuple and EventTypes
    ///////////////////////////////////////////////////////////////////////////

    protected internal sealed class EventCodeTupleSink {
      internal EventCodeTuple eventCodeTuple;
      internal EventType[] eventTypes;
      internal void clear() {
        eventCodeTuple = null;
        eventTypes = null;
      }
    }

    protected internal void createEventCodeTuple(List<EventType> schemaEventTypes, short grammarOptions, EventCodeTupleSink @out, EventTypeList eventTypeList) {
      createEventCodeTuple(schemaEventTypes, grammarOptions, @out, (List<EventType>)null, eventTypeList, false, EXISchema.NIL_GRAM, false, -1, -1);
    }

    protected internal void createEventCodeTuple(List<EventType> schemaEventTypes, short grammarOptions, EventCodeTupleSink @out, List<EventType> invalidAttributes, EventTypeList eventTypeList, bool atZero, int nd, bool isElem, int gramTypeEmpty, int gramContent) {
      int gram, elem;
      bool isEmptyGrammar;
      if (isElem) {
        elem = nd; // nd represents an elem when isElem is true
        Debug.Assert(elem != EXISchema.NIL_NODE);
        int tp = schema.getTypeOfElem(nd);
        gram = schema.getGrammarOfType(tp);
        Debug.Assert(gram != EXISchema.NIL_GRAM);
        isEmptyGrammar = false;
      }
      else {
        elem = EXISchema.NIL_NODE;
        gram = nd; // nd represents a gram when isElem is false;
        bool isTypeGrammar = gram != EXISchema.NIL_GRAM;
        if (isTypeGrammar) {
          isEmptyGrammar = gram == gramTypeEmpty;
        }
        else {
          isEmptyGrammar = false;
        }
      }
      if (gramContent == EXISchema.NIL_GRAM) {
        gramContent = gram;
      }
      int n_invalidAttributes = invalidAttributes != null ? invalidAttributes.Count : 0;

      bool addUndeclaredEA = GrammarOptions.isPermitDeviation(grammarOptions);
      Debug.Assert(invalidAttributes == null || addUndeclaredEA); // invalidAttributes is non-null only when addUndeclaredEA is true
      bool addDTD = GrammarOptions.hasDTD(grammarOptions);
      bool addCM = GrammarOptions.hasCM(grammarOptions);
      bool addPI = GrammarOptions.hasPI(grammarOptions);
      bool addNS, addSC, nillable, typable;

      if (atZero) {
        if (grammarType == SCHEMA_GRAMMAR_ELEMENT_FRAGMENT) {
          typable = nillable = true;
        }
        else {
          Debug.Assert(grammarType == SCHEMA_GRAMMAR_ELEMENT_AND_TYPE);
          if (isElem) {
            bool nilTypeRestricted = GrammarOptions.isXsiNilTypeRestricted(grammarOptions);
            typable = !nilTypeRestricted || schema.isTypableType(schema.getTypeOfElem(elem));
            nillable = !nilTypeRestricted || schema.isNillableElement(elem);
          }
          else {
            bool nillableTypable = !GrammarOptions.isXsiNilTypeRestricted(grammarOptions);
            typable = nillableTypable;
            nillable = nillableTypable;
          }
        }
        addNS = GrammarOptions.hasNS(grammarOptions);
        addSC = GrammarOptions.hasSC(grammarOptions);
      }
      else {
        typable = nillable = addNS = addSC = false;
      }

      bool addUndeclaredEE = false;
      EventType undeclaredEE = null;
      EventType eventTypeSchemaType = null;
      EventType eventTypeSchemaNil = null;
      EventType eventTypeNS = null;
      EventType eventTypeSC = null;
      EventType undeclaredWildcardAnyAT = null;
      EventType undeclaredWildcardAnyInvalidAT = null;
      EventType elementWildcard = null;
      EventType untypedCharacters = null;
      EventType entityReference = null;
      EventType comment = null;
      EventType processingInstruction = null;
      bool addTupleL2 = false;
      bool addTupleL3a = false;
      bool addTupleL3b = false;
      int n_itemsL2 = 0;
      int n_itemsL2Null = 0; // To count how many items in tupleL2 are null.
      int n_itemsL3b = 0;
      int i, len;
      len = schemaEventTypes.Count;
      if (addUndeclaredEA) {
        for (i = 0; i < len; i++) {
          EventType eventType = schemaEventTypes[i];
          if (eventType.itemType == EventType.ITEM_EE) {
            break;
          }
        }
        if (i == len) {
          addUndeclaredEE = true;
          undeclaredEE = EventTypeFactory.creatEndElement(EventCode.EVENT_CODE_DEPTH_TWO, eventTypeList);
          ++n_itemsL2;
        }
        elementWildcard = new EventType(EventCode.EVENT_CODE_DEPTH_TWO, eventTypeList, EventType.ITEM_SE_WC, retrieveEXIGrammar(gramContent));
        untypedCharacters = new EventType(EventCode.EVENT_CODE_DEPTH_TWO, eventTypeList, EventType.ITEM_CH, retrieveEXIGrammar(gramContent));
        n_itemsL2 += 2;
        if (invalidAttributes != null) {
          undeclaredWildcardAnyAT = new EventType(EventCode.EVENT_CODE_DEPTH_TWO, eventTypeList, EventType.ITEM_SCHEMA_AT_WC_ANY, (IGrammar)null);
          undeclaredWildcardAnyInvalidAT = new EventType(EventCode.EVENT_CODE_DEPTH_THREE, eventTypeList, EventType.ITEM_AT_WC_ANY_UNTYPED, (IGrammar)null);
          n_itemsL2 += 2;
          addTupleL3a = true;
        }
        addTupleL2 = true;
      }
      if (typable) {
        if (!isEmptyGrammar) {
          eventTypeSchemaType = createEventTypeXsiType(eventTypeList);
        }
        else {
          ++n_itemsL2Null;
        }
        ++n_itemsL2;
        addTupleL2 = true;
      }
      if (nillable) {
        if (!isEmptyGrammar) {
          eventTypeSchemaNil = createEventTypeXsiNil(eventTypeList, retrieveEXIGrammar(gramTypeEmpty));
        }
        else {
          ++n_itemsL2Null;
        }
        ++n_itemsL2;
        addTupleL2 = true;
      }
      if (addNS) {
        eventTypeNS = new EventType(EventCode.EVENT_CODE_DEPTH_TWO, eventTypeList, EventType.ITEM_NS, (IGrammar)null);
        ++n_itemsL2;
        addTupleL2 = true;
      }
      if (addSC) {
        eventTypeSC = new EventType(EventCode.EVENT_CODE_DEPTH_TWO, eventTypeList, EventType.ITEM_SC, (IGrammar)null);
        ++n_itemsL2;
        addTupleL2 = true;
      }
      if (addDTD) {
        entityReference = new EventType(EventCode.EVENT_CODE_DEPTH_TWO, eventTypeList, EventType.ITEM_ER, retrieveEXIGrammar(gramContent));
        ++n_itemsL2;
        addTupleL2 = true;
      }
      if (addCM) {
        comment = new EventType(EventCode.EVENT_CODE_DEPTH_THREE, eventTypeList, EventType.ITEM_CM, retrieveEXIGrammar(gramContent));
        ++n_itemsL3b;
        addTupleL3b = true;
      }
      if (addPI) {
        processingInstruction = new EventType(EventCode.EVENT_CODE_DEPTH_THREE, eventTypeList, EventType.ITEM_PI, retrieveEXIGrammar(gramContent));
        ++n_itemsL3b;
        addTupleL3b = true;
      }
      if (addTupleL3b) {
        ++n_itemsL2;
        addTupleL2 = true;
      }
      ArrayEventCodeTuple eventCodes;
      EventCode[] eventCodeItems;
      EventCode[] eventCodeItemsL2 = null;
      EventCode[] eventCodeItemsL3a = null;
      EventCode[] eventCodeItemsL3b = null;
      ArrayEventCodeTuple tupleL2 = null;
      ArrayEventCodeTuple tupleL3a = null;
      ArrayEventCodeTuple tupleL3b = null;
      int totalLen = n_itemsL3b != 0 ? len + ((n_itemsL2 - n_itemsL2Null) - 1) + n_itemsL3b + n_invalidAttributes: len + (n_itemsL2 - n_itemsL2Null) + n_invalidAttributes;
      eventCodes = new ArrayEventCodeTuple();
      EventType[] eventTypes = new EventType[totalLen];
      eventCodeItems = new EventCode[addTupleL2 ? len + 1 : len];
      int n = 0;
      if (eventTypeSchemaType != null) {
        eventTypes[n++] = eventTypeSchemaType;
      }
      if (eventTypeSchemaNil != null) {
        eventTypes[n++] = eventTypeSchemaNil;
      }
      for (i = 0; i < len; i++) {
        EventType eventType = schemaEventTypes[i];
        eventTypes[n + i] = eventType;
        eventCodeItems[i] = eventType;
      }
      n += len;
      if (addTupleL2) {
        eventCodeItemsL2 = new EventCode[n_itemsL2];
        tupleL2 = new ArrayEventCodeTuple();
        eventCodeItems[len] = tupleL2;
        if (addTupleL3b) {
          eventCodeItemsL3b = new EventCode[n_itemsL3b];
          tupleL3b = new ArrayEventCodeTuple();
          eventCodeItemsL2[n_itemsL2 - 1] = tupleL3b;
        }
      }
      int m = 0, k = 0;
      if (addUndeclaredEE) {
        eventTypes[n++] = undeclaredEE;
        eventCodeItemsL2[m++] = undeclaredEE;
      }
      if (typable) {
        eventCodeItemsL2[m++] = eventTypeSchemaType;
      }
      if (nillable) {
        eventCodeItemsL2[m++] = eventTypeSchemaNil;
      }
      if (addTupleL3a) {
        eventTypes[n++] = undeclaredWildcardAnyAT;
        eventCodeItemsL2[m++] = undeclaredWildcardAnyAT;
        eventCodeItemsL3a = new EventCode[n_invalidAttributes + 1];
        tupleL3a = new ArrayEventCodeTuple();
        eventCodeItemsL2[m++] = tupleL3a;
        for (i = 0; i < n_invalidAttributes; i++) {
          eventTypes[n++] = invalidAttributes[i];
          eventCodeItemsL3a[i] = invalidAttributes[i];
        }
        eventTypes[n++] = undeclaredWildcardAnyInvalidAT;
        eventCodeItemsL3a[n_invalidAttributes] = undeclaredWildcardAnyInvalidAT;
      }
      if (addNS) {
        eventTypes[n++] = eventTypeNS;
        eventCodeItemsL2[m++] = eventTypeNS;
      }
      if (addSC) {
        eventTypes[n++] = eventTypeSC;
        eventCodeItemsL2[m++] = eventTypeSC;
      }
      if (addUndeclaredEA) {
        eventTypes[n++] = elementWildcard;
        eventCodeItemsL2[m++] = elementWildcard;
        eventTypes[n++] = untypedCharacters;
        eventCodeItemsL2[m++] = untypedCharacters;
      }
      if (addDTD) {
        eventTypes[n++] = entityReference;
        eventCodeItemsL2[m++] = entityReference;
      }
      if (addCM) {
        eventTypes[n++] = comment;
        eventCodeItemsL3b[k++] = comment;
      }
      if (addPI) {
        eventTypes[n++] = processingInstruction;
        eventCodeItemsL3b[k++] = processingInstruction;
      }
      Debug.Assert(totalLen == n && (addTupleL3b ? n_itemsL2 - 1: n_itemsL2) == m && n_itemsL3b == k);
      eventCodes.Items = eventCodeItems;
      if (addTupleL2) {
        tupleL2.Items = eventCodeItemsL2;
        if (addTupleL3a) {
          tupleL3a.Items = eventCodeItemsL3a;
        }
        if (addTupleL3b) {
          tupleL3b.Items = eventCodeItemsL3b;
        }
      }

      @out.eventCodeTuple = eventCodes;
      @out.eventTypes = eventTypes;
    }

  }

}

File: fujitsu/src/org/openexi/fujitsu/proc/grammars/SchemaInformedGrammar.java
package org.openexi.fujitsu.proc.grammars;

import java.util.ArrayList;

import org.openexi.fujitsu.proc.common.EventCode;
import org.openexi.fujitsu.proc.common.EventType;
import org.openexi.fujitsu.proc.common.EventTypeList;
import org.openexi.fujitsu.proc.common.GrammarOptions;
import org.openexi.fujitsu.schema.EXISchema;

public abstract class SchemaInformedGrammar extends Grammar {

  protected static final byte ELEMENT_FRAGMENT_STATE_BASE = DOCUMENT_STATE_END + 1;
  public static final byte ELEMENT_FRAGMENT_STATE_TAG           = ELEMENT_FRAGMENT_STATE_BASE;
  public static final byte ELEMENT_FRAGMENT_STATE_CONTENT       = ELEMENT_FRAGMENT_STATE_TAG + 1;
  public static final byte ELEMENT_FRAGMENT_EMPTY_STATE_TAG     = ELEMENT_FRAGMENT_STATE_CONTENT + 1;
  public static final byte ELEMENT_FRAGMENT_EMPTY_STATE_CONTENT = ELEMENT_FRAGMENT_EMPTY_STATE_TAG + 1;
  public static final byte ELEMENT_FRAGMENT_BOUND               = ELEMENT_FRAGMENT_EMPTY_STATE_CONTENT + 1;

  private static final byte ELEMENT_STATE_BASE = ELEMENT_FRAGMENT_BOUND + 1;
  // SchemaInformedElementGrammar
  public static final byte ELEMENT_STATE_UNBOUND  = ELEMENT_STATE_BASE;
  public static final byte ELEMENT_STATE_BOUND    = ELEMENT_STATE_UNBOUND + 1;
  // ElementTagGrammar
  public static final byte ELEMENT_STATE_TAG     = ELEMENT_STATE_BOUND + 1;
  protected static final byte ELEMENT_STATE_CONTENT = ELEMENT_STATE_TAG + 1;

  private static final byte ELEMENT_STATE_CONTENT_BASE = ELEMENT_STATE_CONTENT + 1;
  // ElementContentGrammar
  protected static final byte ELEMENT_STATE_CONTENT_DEPLETE     = ELEMENT_STATE_CONTENT_BASE;
  protected static final byte ELEMENT_STATE_CONTENT_ACCEPTED    = ELEMENT_STATE_CONTENT_DEPLETE + 1;
  public static final byte ELEMENT_STATE_CONTENT_COMPLETE    = ELEMENT_STATE_CONTENT_ACCEPTED + 1;


  protected final int[] m_nodes;
  protected final String[] m_names;

  /**
  * schema node (if STATE_DOCUMENT), element node (if STATE_ELEMENT),
  * group node (if STATE_GROUP)
  */
  final int   m_nd;
  
  protected SchemaInformedGrammar(int nd, byte grammarType, GrammarCache stateCache) {
    super(grammarType, stateCache);
    m_nd = nd;
    if (m_schema != null) {
      m_nodes  = m_schema.getNodes();
      m_names  = m_schema.getNames();
    }
    else {
      m_nodes  = null;
      m_names  = null;
    }
  }

  @Override
  public final boolean isSchemaInformed() {
    return true;
  }

  final int getNode() {
    return m_nd;
  }
  
  abstract void schemaAttribute(int eventTypeIndex, String uri, String name, GrammarState stateVariables);

  protected final boolean schemaElement(final EventTypeSchema schemaEventType, String uri, String name, GrammarState stateVariables) {
    EventTypeList eventTypeList = getNextEventTypes(stateVariables);
    for (int i = 0; i < eventTypeList.getLength(); i++) {
      EventType ith = eventTypeList.item(i);
      if (ith.isSchemaInformed() && schemaEventType.getSchemaSubstance() == ((EventTypeSchema)ith).getSchemaSubstance()) {
        element(i, uri, name, stateVariables);
        return true;
      }
    }
    return false;
  }

  /**
   * Sort particles in EXI schema order.
   * @param initials the location where initials start in m_nodes
   * @param the number of initials
   */
  protected void sortInitials(int[] sortedInitials, int initials, int n_initials) {
    assert sortedInitials.length == n_initials; // caller is responsible for allocation
    assert initials != EXISchema.NIL_NODE;
    int i;
    for (i = 0; i < n_initials; i++) {
      sortedInitials[i] = i;
    }
    for (i = 0; i < n_initials - 1; i++) {
      for (int j = i + 1; j < n_initials; j++) {
        final int particle_j;
        if ((particle_j = m_nodes[initials + sortedInitials[j]]) == EXISchema.NIL_NODE) {
          assert j == n_initials - 1;
          break;
        }
        final int particle_i, term_i, sn_i;
        particle_i = m_nodes[initials + sortedInitials[i]];
        assert particle_i != EXISchema.NIL_NODE;
        term_i = m_schema.getTermOfParticle(particle_i);
        sn_i = m_schema.getSerialInTypeOfParticle(particle_i);

        final int term_j, sn_j;
        term_j = m_schema.getTermOfParticle(particle_j);
        sn_j = m_schema.getSerialInTypeOfParticle(particle_j);
        
        if (!isAscending(particle_i, term_i, sn_i, particle_j, term_j, sn_j)) {
          final int _saved;
          _saved = sortedInitials[i];
          sortedInitials[i] =  sortedInitials[j];
          sortedInitials[j] = _saved;
        }
      }
    }
  }
  
  /**
   * Determines if (particle1, term1, sn1) <= (particle2, term2, sn2) in EXI schema order.
   */
  protected boolean isAscending(int particle1, int term1, int sn1, int particle2, int term2, int sn2) {
    assert (sn1 < 0 || particle1 != EXISchema.NIL_NODE && term1 != EXISchema.NIL_NODE) &&
      (sn2 == Integer.MAX_VALUE || particle2 != EXISchema.NIL_NODE && term2 != EXISchema.NIL_NODE);
    if (sn1 < 0) {
      assert particle1 == EXISchema.NIL_NODE && term1 == EXISchema.NIL_NODE && sn2 >= 0;
      return true;
    }
    if (sn2 == Integer.MAX_VALUE) {
      assert particle2 == EXISchema.NIL_NODE && term2 == EXISchema.NIL_NODE && sn1 < sn2;
      return true;
    }
    final int termType1, termType2;
    termType1 = m_nodes[term1];
    termType2 = m_nodes[term2];
    
    if (termType1 == EXISchema.ELEMENT_NODE && termType2 == EXISchema.WILDCARD_NODE) {
      return true;
    }
    else if (termType1 == EXISchema.WILDCARD_NODE && termType2 == EXISchema.ELEMENT_NODE) {
      return false;
    }
    
    if (termType1 == EXISchema.ELEMENT_NODE) {
      assert termType2 == EXISchema.ELEMENT_NODE;
      if (sn1 <= sn2) {
        assert sn1 < sn2 || particle1 == particle2 && term1 != term2;
        return true;
      }
    }
    else {
      assert termType1 == EXISchema.WILDCARD_NODE && termType2 == EXISchema.WILDCARD_NODE;
      int constraint1, constraint2;
      constraint1 = m_schema.getConstraintTypeOfWildcard(term1);
      constraint2 = m_schema.getConstraintTypeOfWildcard(term2);
      if (constraint1 != EXISchema.WC_TYPE_NAMESPACES)
        constraint1 = EXISchema.WC_TYPE_ANY;
      if (constraint2 != EXISchema.WC_TYPE_NAMESPACES)
        constraint2 = EXISchema.WC_TYPE_ANY;
      assert constraint1 != constraint2 || constraint1 == EXISchema.WC_TYPE_NAMESPACES;
      if (constraint1 == EXISchema.WC_TYPE_NAMESPACES && constraint2 == EXISchema.WC_TYPE_ANY)
        return true;
      else if (constraint1 == EXISchema.WC_TYPE_ANY && constraint2 == EXISchema.WC_TYPE_NAMESPACES)
        return false;
      assert constraint1 == EXISchema.WC_TYPE_NAMESPACES && constraint2 == EXISchema.WC_TYPE_NAMESPACES;
      if (sn1 <= sn2)
        return true;
    }
    return false;
  }
  
  ///////////////////////////////////////////////////////////////////////////
  /// Methods computing EventCodeTuple and EventTypes
  ///////////////////////////////////////////////////////////////////////////

  protected static final class EventCodeTupleSink {
    EventCodeTuple eventCodeTuple;
    AbstractEventType[] eventTypes;
    void clear() {
      eventCodeTuple = null;
      eventTypes = null;
    }
  }

  protected final void createEventCodeTuple(final ArrayList<AbstractEventType> schemaEventTypes, 
      final short grammarOptions, final EventCodeTupleSink out, EventTypeList eventTypeList) {
    createEventCodeTuple(schemaEventTypes, grammarOptions, out, 
        (ArrayList<EventTypeSchemaAttributeInvalid>)null, eventTypeList, false);
  }

  protected final void createEventCodeTuple(final ArrayList<AbstractEventType> schemaEventTypes, 
      final short grammarOptions, final EventCodeTupleSink out, 
      final ArrayList<EventTypeSchemaAttributeInvalid> invalidAttributes, EventTypeList eventTypeList, final boolean atZero) {
    
    final int n_invalidAttributes = invalidAttributes != null ? invalidAttributes.size() : 0;
    
    final boolean addUndeclaredEA = GrammarOptions.hasUndeclaredEA(grammarOptions);
    assert invalidAttributes == null || addUndeclaredEA; // invalidAttributes is non-null only when addUndeclaredEA is true
    final boolean addDTD = GrammarOptions.hasDTD(grammarOptions);
    final boolean addCM =  GrammarOptions.hasCM(grammarOptions);
    final boolean addPI =  GrammarOptions.hasPI(grammarOptions);
    final boolean addNS, addSC, nillable, typable;
    
    if (atZero) {
      if (m_grammarType == SCHEMA_GRAMMAR_ELEMENT) {
        if (m_nd != EXISchema.NIL_NODE) {
          final int tp;
          final boolean nilTypeRestricted = GrammarOptions.isXsiNilTypeRestricted(grammarOptions);
          tp = m_schema.getTypeOfElem(m_nd);
          typable = !nilTypeRestricted || m_schema.hasSubType(tp) || 
            m_schema.getNodeType(tp) == EXISchema.SIMPLE_TYPE_NODE && m_schema.getVarietyOfSimpleType(tp) == EXISchema.UNION_SIMPLE_TYPE;
          nillable = !nilTypeRestricted || m_schema.isNillableElement(m_nd);
        }
        else {
          typable = true;
          nillable = true;
        }
      }
      else if (m_grammarType == SCHEMA_GRAMMAR_ELEMENT_TAG) {
        final boolean nillableTypable = !GrammarOptions.isXsiNilTypeRestricted(grammarOptions);
        typable  = nillableTypable;
        nillable = nillableTypable;
      }
      else {
        assert m_grammarType == SCHEMA_GRAMMAR_ELEMENT_FRAGMENT;
        typable  = nillable = true;
      }
      addNS = GrammarOptions.hasNS(grammarOptions);
      addSC = GrammarOptions.hasSC(grammarOptions);
    }
    else {
      typable = nillable = addNS = addSC = false;
    }
    
    boolean addUndeclaredEE = false;
    EventTypeEndElementSecond undeclaredEE = null;  
    EventTypeSchemaType eventTypeSchemaType = null;
    EventTypeSchemaNil eventTypeSchemaNil = null;
    EventTypeNamespaceDeclaration eventTypeNS = null;
    EventTypeSelfContained eventTypeSC = null;
    EventTypeSchemaAttributeWildcardAny undeclaredWildcardAnyAT = null;
    EventTypeAttributeWildcardAnyUntyped undeclaredWildcardAnyInvalidAT = null;
    EventTypeElementWildcard elementWildcard = null;
    EventTypeCharacters untypedCharacters = null;
    EventTypeEntityReference entityReference = null;
    EventTypeComment comment = null;
    EventTypeProcessingInstruction processingInstruction = null;
    boolean addTupleL2 = false;
    boolean addTupleL3a = false;
    boolean addTupleL3b = false;
    int n_itemsL2 = 0;
    int n_itemsL3b = 0;
    int i,  len;
    len = schemaEventTypes.size();
    if (addUndeclaredEA) {
      for (i = 0; i < len; i++) {
        final EventType eventType = schemaEventTypes.get(i);
        if (eventType.itemType == EventCode.ITEM_SCHEMA_EE)
          break;
      }
      if (i == len) {
        addUndeclaredEE = true;
        undeclaredEE = new EventTypeEndElementSecond(this, eventTypeList); 
        ++n_itemsL2;
      }
      elementWildcard = new EventTypeElementWildcard(EventCode.EVENT_CODE_DEPTH_TWO, this, eventTypeList);
      untypedCharacters = new EventTypeCharactersSecond(EventCode.EVENT_CODE_DEPTH_TWO, this, eventTypeList);
      n_itemsL2 += 2;
      if (invalidAttributes != null) {
        undeclaredWildcardAnyAT = EventTypeSchemaAttributeWildcardAny.createLevelTwo(this, eventTypeList);
        undeclaredWildcardAnyInvalidAT = new EventTypeAttributeWildcardAnyUntyped(EventCode.EVENT_CODE_DEPTH_THREE, this, eventTypeList); 
        n_itemsL2 += 2;
        addTupleL3a = true;
      }
      addTupleL2 = true;
    }
    if (typable) {
      eventTypeSchemaType = new EventTypeSchemaType(this, eventTypeList);
      ++n_itemsL2;
      addTupleL2 = true;
    }
    if (nillable) {
      eventTypeSchemaNil = new EventTypeSchemaNil(this, eventTypeList);
      ++n_itemsL2;
      addTupleL2 = true;
    }
    if (addNS) {
      eventTypeNS = new EventTypeNamespaceDeclaration(this, eventTypeList);
      ++n_itemsL2;
      addTupleL2 = true;
    }
    if (addSC) {
      eventTypeSC = new EventTypeSelfContained(this, eventTypeList);
      ++n_itemsL2;
      addTupleL2 = true;
    }
    if (addDTD) {
      entityReference = new EventTypeEntityReference(this, eventTypeList);
      ++n_itemsL2;
      addTupleL2 = true;
    }
    if (addCM) {
      comment = new EventTypeComment(EventCode.EVENT_CODE_DEPTH_THREE, this, eventTypeList);
      ++n_itemsL3b;
      addTupleL3b = true;
    }
    if (addPI) {
      processingInstruction = new EventTypeProcessingInstruction(EventCode.EVENT_CODE_DEPTH_THREE, this, eventTypeList);
      ++n_itemsL3b;
      addTupleL3b = true;
    }
    if (addTupleL3b) {
      ++n_itemsL2;
      addTupleL2 = true;
    }
    ArrayEventCodeTuple eventCodes;
    EventCode[] eventCodeItems;
    EventCode[] eventCodeItemsL2 = null;
    EventCode[] eventCodeItemsL3a = null;
    EventCode[] eventCodeItemsL3b = null;
    ArrayEventCodeTuple tupleL2 = null;
    ArrayEventCodeTuple tupleL3a = null;
    ArrayEventCodeTuple tupleL3b = null;
    final int totalLen = n_itemsL3b != 0 ? 
        len + (n_itemsL2 - 1) + n_itemsL3b + n_invalidAttributes: len + n_itemsL2 + n_invalidAttributes;
    eventCodes = ArrayEventCodeTuple.createTuple(); 
    AbstractEventType[] eventTypes = new AbstractEventType[totalLen];
    eventCodeItems = new EventCode[addTupleL2 ? len + 1 : len];
    int n = 0;
    if (typable) {
      eventTypes[n++] = eventTypeSchemaType;
    }
    if (nillable) {
      eventTypes[n++] = eventTypeSchemaNil;
    }
    for (i = 0; i < len; i++) {
      final AbstractEventType eventType = schemaEventTypes.get(i);
      eventTypes[n + i] = eventType;
      eventCodeItems[i] = eventType;
    }
    n += len;
    if (addTupleL2) {
      eventCodeItemsL2 = new EventCode[n_itemsL2];
      tupleL2 = ArrayEventCodeTuple.createTuple();
      eventCodeItems[len] = tupleL2;
      if (addTupleL3b) {
        eventCodeItemsL3b = new EventCode[n_itemsL3b];
        tupleL3b = ArrayEventCodeTuple.createTuple();
        eventCodeItemsL2[n_itemsL2 - 1] = tupleL3b;
      }
    }
    int m = 0, k = 0;
    if (addUndeclaredEE) {
      eventTypes[n++] = undeclaredEE;
      eventCodeItemsL2[m++] = undeclaredEE;
    }
    if (typable) {
      eventCodeItemsL2[m++] = eventTypeSchemaType;
    }
    if (nillable) {
      eventCodeItemsL2[m++] = eventTypeSchemaNil;
    }
    if (addTupleL3a) {
      eventTypes[n++] = undeclaredWildcardAnyAT;
      eventCodeItemsL2[m++] = undeclaredWildcardAnyAT;
      eventCodeItemsL3a =  new EventCode[n_invalidAttributes + 1];
      tupleL3a = ArrayEventCodeTuple.createTuple();
      eventCodeItemsL2[m++] = tupleL3a;
      for (i = 0; i < n_invalidAttributes; i++) {
        eventTypes[n++] = invalidAttributes.get(i);
        eventCodeItemsL3a[i] = invalidAttributes.get(i);
      }
      eventTypes[n++] = undeclaredWildcardAnyInvalidAT;
      eventCodeItemsL3a[n_invalidAttributes] = undeclaredWildcardAnyInvalidAT;
    }
    if (addNS) {
      eventTypes[n++] = eventTypeNS; 
      eventCodeItemsL2[m++] = eventTypeNS; 
    }
    if (addSC) {
      eventTypes[n++] = eventTypeSC; 
      eventCodeItemsL2[m++] = eventTypeSC; 
    }
    if (addUndeclaredEA) {
      eventTypes[n++] = elementWildcard;
      eventCodeItemsL2[m++] = elementWildcard;
      eventTypes[n++] = untypedCharacters;
      eventCodeItemsL2[m++] = untypedCharacters;
    }
    if (addDTD) {
      eventTypes[n++] = entityReference;
      eventCodeItemsL2[m++] = entityReference;
    }
    if (addCM) {
      eventTypes[n++] = comment;
      eventCodeItemsL3b[k++] = comment;
    }
    if (addPI) {
      eventTypes[n++] = processingInstruction;
      eventCodeItemsL3b[k++] = processingInstruction;
    }
    assert totalLen == n && (addTupleL3b ? n_itemsL2 - 1 : n_itemsL2) == m && n_itemsL3b == k;
    eventCodes.setItems(eventCodeItems);
    if (addTupleL2) {
      tupleL2.setItems(eventCodeItemsL2);
      if (addTupleL3a) {
        tupleL3a.setItems(eventCodeItemsL3a);
      }
      if (addTupleL3b) {
        tupleL3b.setItems(eventCodeItemsL3b);
      }
    }
    
    out.eventCodeTuple = eventCodes;
    out.eventTypes = eventTypes;
  }
  
}
